File: NURBS_viewer/autogl/sample_f/pick2d.f
c     
c     メインルーチン
c     
      implicit none

      real*8 vertexs(6, 2)
      integer selected_vertex_id
      common vertexs, selected_vertex_id

c     頂点の座標 
      vertexs(1, 1) = 10.0d0
      vertexs(1, 2) = 0.0d0

      vertexs(2, 1) = 10.0d0
      vertexs(2, 2) = 20.0d0

      vertexs(3, 1) = 0.0d0
      vertexs(3, 2) = 20.0d0

      vertexs(4, 1) = -10.0d0
      vertexs(4, 2) = 0.0d0

      vertexs(5, 1) = 0.0d0
      vertexs(5, 2) = -20.0d0

      vertexs(6, 1) = 30.0d0
      vertexs(6, 2) = -20.0d0

c     現在選択されている頂点の番号 
      selected_vertex_id = -1



      call autogl_set_view_size(70.0d0)

c     イベント処理コールバック関数としてautogl_handle_eventを指定。 
      call autogl_set_default_callback_in_mode2d()

c     初期状態では、ユーザー定義イベント処理を行う。 
      call autogl_set_mode2d(0)

c     イベント処理中, ドラッグを有効にする。
      call autogl_enable_drag_in_mode2d()

c     パネル上に二次元アプリ用イベント処理機能を準備する。 
      call autogl_set_panel_in_mode2d()

c     AutoGLのメインループ
c     二度と戻らない
      call autogl_main()

      end



c     サブルーチンautoglredraw
c     ビューの再描画のためのコールバックサブルーチン 
c     ビューが再表示されるごとに呼ばれる。 
      subroutine autogl_redraw
      implicit none

      real*8 vertexs(6, 2)
      integer selected_vertex_id
      common vertexs, selected_vertex_id

      integer i
      integer mark_size_dc
      integer is_filled

      call autogl_set_color(0.0d0, 1.0d0, 0.0d0) 

      call autogl_draw_line(
     c     -50.0d0, 0.0d0, 0.0d0,
     c     50.0d0, 0.0d0, 0.0d0)
      call autogl_draw_string(50.0d0, 0.0d0, 0.0d0, "X = 50") 
      call autogl_draw_string(-50.0d0, 0.0d0, 0.0d0, "X = -50") 
         
      call autogl_draw_line(
     c     0.0d0, -50.0d0, 0.0d0,
     c     0.0d0, 50.0d0, 0.0d0)
      call autogl_draw_string(0.0d0, 50.0d0, 0.0d0, "Y = 50") 
      call autogl_draw_string(0.0d0, -50.0d0, 0.0d0, "Y = -50") 

      call autogl_set_color(1.0d0, 1.0d0, 1.0d0) 
      call autogl_draw_string(0.0d0, 0.0d0, 0.0d0, "O")

      call autogl_set_color(1.0d0, 1.0d0, 1.0d0) 
      do i = 1, 6 - 1
         call autogl_draw_line(
     c        vertexs(i, 1), vertexs(i, 2), 0.0d0,
     c        vertexs(i + 1, 1), vertexs(i + 1, 2), 0.0d0)
      enddo
      
c     各頂点について、それが選択されているかどうかを描画する。 
      call autogl_set_color(1.0d0, 0.0d0, 0.0d0) 
      do i = 1, 6
         is_filled = 0
         if (i .eq. selected_vertex_id) then
            is_filled = 1
         endif
         
         mark_size_dc = 10
c     ダイアモンドのマークを表示する。 
         call autogl_draw_diamond_mark2d (
     c        vertexs(i, 1), vertexs(i, 2), 0.0d0,
     c        mark_size_dc, is_filled)
c     mark_size_dcにより、10 pixelぐらいの大きさとなる。 
c     is_filled .eq. 1なら、この頂点は選択されている。 
      enddo

      end



      subroutine autogl_handle_event
      implicit none

      real*8 vertexs(6, 2)
      integer selected_vertex_id
      common vertexs, selected_vertex_id

      integer event
      integer i
      integer tolerance_dc
      integer flag
      
      call autogl_get_view_event(event)

c     もし、マウスクリックがあれば 
      if (event .eq. 2) then 
         
c     各頂点について、マウスポインタがヒットしたかどうか調べる。 
         selected_vertex_id = -1
         do i = 1, 6
            tolerance_dc = 10
            
c     この頂点がマウスポインタのそばにあれば 
            call autogl_pointing_device_is_hit2d(
     c           flag, 
     c           vertexs(i, 1), vertexs(i, 2),
     c           tolerance_dc)
c     点の座標は世界座標（ただし二次元）で与える。 
            if (flag .eq. 1) then
               selected_vertex_id = i
            endif
         enddo
         
         if (selected_vertex_id .ne. -1) then
            write(*,*) ' vertex ', selected_vertex_id, ' is selected '
         endif

c     ビューを再描画する。 
         call autogl_draw_view()

c     もし、キーボード入力があれば 
      elseif (event .eq. 3) then
         call autogl_handle_default_key_event_in_mode2d()
      endif

      end



      subroutine autogl_idle
      end

      subroutine autogl_callback(id)
      implicit none
      integer id
      end

File: NURBS_viewer/autogl/core/autogl_mp/sample_c/pick2d.c
/*  AutoGL ＜お手軽＞ のサンプルプログラム例 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>

/* AutoGL ”おてがる”を使うには,このヘッダファイルをインクルードする。*/
#include <autogl.h>

/* ユーティリティの二次元イベント処理の利用 */
/* 引数など詳しい情報は, autogl_utility2d.hを参照。*/
/* なお、これはautogl.hからincludeされている。*/



/* ビュー上に表示されるモデルを表現するための変数群 */

/* 現在選択されている頂点の番号 */
static int SelectedVertexId = -1;

/* 頂点の座標 */
static double Vertexs[6][2] = {
  { 10, 0 },
  { 10, 20 },
  { 0, 20 },
  { -10, 0 },
  { 0, -20 },
  { 30, -20 }
};



/* ビューの再描画のためのコールバック関数 */
/* ビューが再表示されるごとに呼ばれる。 */
static void RedrawView (void) 
{
  int i;
  int isFilled;

  AutoGL_SetColor (0, 1, 0);  

  AutoGL_DrawLine (-50, 0, 0,
		   50, 0, 0); 
  AutoGL_DrawString (50, 0, 0, "X = 5O");  
  AutoGL_DrawString (-50, 0, 0, "X = -5O");  
  
  AutoGL_DrawLine (0, -50, 0,
		   0, 50, 0); 
  AutoGL_DrawString (0, 50, 0, "Y = 5O");  
  AutoGL_DrawString (0, -50, 0, "Y = -5O");  
    
  AutoGL_SetColor (1, 1, 1);
  AutoGL_DrawString (0, 0, 0, "O"); 

  AutoGL_SetColor (1, 1, 1);
  for (i = 0; i < 6 - 1; i++) {
    AutoGL_DrawLine (Vertexs[i][0], Vertexs[i][1], 0,
		     Vertexs[i + 1][0], Vertexs[i + 1][1], 0);
  }

  /* 各頂点について、それが選択されているかどうかを描画する。 */
  AutoGL_SetColor (1, 0, 0);
  for (i = 0; i < 6; i++) {
    int markSizeDc = 10;

    isFilled = 0;
    if (i == SelectedVertexId) {
      isFilled = 1;
    }

    /* ダイアモンドのマークを表示する。 */
    AutoGL_DrawDiamondMark2D 
      (Vertexs[i][0], Vertexs[i][1], 0, 
       markSizeDc, isFilled);
    /* markSizeDcにより、10 pixelぐらいの大きさとなる。 */
    /* isFilled == 1なら、この頂点は選択されている。 */
  }
}

/* ビューのイベント処理ためのコールバック関数 */
/* ビューにイベントが発生するごとに呼ばれる。 */
static void HandleEvent (void)
{
  int event = AutoGL_GetViewEvent ();

  /* もし、マウスクリックがあれば */
  if (event == AUTOGL_EVENT_POINTING_DEVICE_RELEASE) {
    int i;

    /* 各頂点について、マウスポインタがヒットしたかどうか調べる。 */
    SelectedVertexId = -1;
    for (i = 0; i < 6; i++) {
      int toleranceDc = 10;

      /* この頂点がマウスポインタのそばにあれば */
      if (AutoGL_PointingDeviceIsHit2D 
	  (Vertexs[i][0], Vertexs[i][1], 
	   toleranceDc)) {
	/* 点の座標は世界座標（ただし二次元）で与える。 */
	SelectedVertexId = i;
	break;
      }
    }
    if (SelectedVertexId != -1) {
      printf (" vertex %d is selected. \n",
	      SelectedVertexId);
    }

    /* ビューを再描画する。 */
    AutoGL_DrawView ();
    /* 登録されたビューの再描画関数（この場合はRedrawView ()）が呼ばれる。 */

  } else if (event == AUTOGL_EVENT_KEY) {
    AutoGL_HandleDefaultKeyEventInMode2D ();
  }
}



/* 関数AutoGL_SetUpはユーザー側プログラムごとに必ず一つ用意すること。*/
/* ここで、コールバック関数や制御変数などを登録する。*/
void AutoGL_SetUp (int argc, char *argv[]) 
{
  AutoGL_SetViewSize (70);   

  /* ビューの再描画コールバックの登録 */
  AutoGL_SetViewRedrawCallback (RedrawView);   

  /* ビューのイベント処理コールバックの登録 */
  AutoGL_SetDefaultCallbackInMode2D (HandleEvent);
  /* イベント処理コールバック関数として上にあるHandleEvent ()を指定。 */
  /* 特別なイベント処理が要らなければ、NULLでよい。*/

  /* 初期状態では、ユーザー定義イベント処理を行う。 */
  AutoGL_SetMode2D (AUTOGL_MODE_2D_OTHER); 

#if 0
  /* イベント処理中, ドラッグを有効にする。*/
  AutoGL_EnableDragInMode2D ();
#endif

  /* パネル上に二次元アプリ用イベント処理機能を準備する。 */
  AutoGL_SetPanelInMode2D ();
}

